src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorPage.cs

```diff
@@ -270,6 +270,36 @@ public new bool IsSectionDefined(string name)
             return zone != null && zone.Items.Count > 0;
         }
 
+        public new void DefineSection(string name, Microsoft.AspNetCore.Mvc.Razor.RenderAsyncDelegate section)
+        {
+            var zone = ThemeLayout.Zones[name];
+            dynamic sectionShape = New.AspSection(SectionContent: SectionDelegate(name,section) ).GetAwaiter().GetResult();
+
+            if (zone is Zones.ZoneOnDemand zoneOnDemand)
+            {
+                sectionShape = zoneOnDemand.AddAsync(sectionShape);
+            }
+            else
+            {
+                sectionShape = zone.Add(sectionShape);
+            }
+        }
+
+        private string SectionDelegate(string name, Microsoft.AspNetCore.Mvc.Razor.RenderAsyncDelegate oldDelegate)
+        {
+            string sectionContent = null;
+            Microsoft.AspNetCore.Mvc.Razor.RenderAsyncDelegate newDelegate = async () =>
+            {
+                var temporaryWriter = new System.IO.StringWriter();
+                PushWriter(temporaryWriter);
+                await oldDelegate();
+                PopWriter();
+                sectionContent = temporaryWriter.ToString();
+            };
+            newDelegate().GetAwaiter().GetResult();
```

Comment:
Concern

Reply:
Refactored code to return delegate instead of result. Now RenderAsyncDelegate will be called within  AspSection shape method, when DisplayAsync is called